連堃玹 - 面試

monero778176 opened the pull request:
### Requirements
- Add the "Order" model in api app.
- Create validate token by POST method in "import_order" function.
- Create 3 unittest for token validation in api app.

### Advanced Requirements
- Fix token validation using deractor (/api/view/order.py)
- Add "Product" model, and create the relationship to Order (with fine adjustment)
  -  Add the middle table - "Product Sell" handle between "Product" and "Order".
  - Add list_display on admin, good for browse table for back end

monero778176 committed:
Requirements設定

新增order, product 及其餘 models 設計
POST method 驗證 token

monero778176 committed:
Fix api/admin

修改顯示名稱

monero778176 committed:
Add unittest

Add unittest

WendyYehhh commented on api/models.py:
```diff
@@ -1,8 +1,129 @@
 
 
 # Create your models here.
 class Order(models.Model):
```
一般來說，
- `order_num` 應該要是 unique 的，所以會加上 unique constraint
- `__str__`  是要給使用者看的，應該要用有意義的 `order_num` 取代 資料庫的 `id`

WendyYehhh commented on api/models.py:
```diff
@@ -1,8 +1,129 @@
+    def __str__(self):
+        return self.name
+
+class ProductSell(models.Model):
+    # Foreignkey
+    
+    # 單一一次銷售對應一個商品項目
+    product = models.ForeignKey(Product,on_delete=models.CASCADE, related_name='productSell')
+    order = models.ForeignKey(Order,on_delete=models.CASCADE,related_name='productSell')
+
+    sell_num = models.IntegerField(verbose_name='銷量')
+
+
+    class Meta:
+        db_table='productsell'
```
一般來說這種 through table 會需要加上 unique constraint 
```
constraints = [models.UniqueConstraint(fields=["product", "order"], name="XXX]
```

WendyYehhh reviewed (commented):
- 有文件說明(`.md`) 是很好的！但是現在這樣不太容易閱讀，我會覺得沒有幫助閱讀的人理解。單純看 models.py 可能還比較快且正確。因為你定義了不少 models，如果畫個 ER diagram 會幫助其他人更好理解。
- `models.py` 裡面
  - 有很多看起來比較沒意義的 comments ，最好移除掉。
  - 以及有些多餘的空行也可以移除，比較好閱讀。
- 因為你加上了比 requirements 更多 models ，會比較希望也一並實作 相關 api endpoints，不然單純定義這些 models 比較沒意義，也難以驗證 schema 是否符合商業需求 (這裡不一定指的是現實商業中的需求，因為你可能也沒碰過，可以是你在 views 自己定義允許的操作行為，至少有這些基本的流程，我們才能知道這些 models 設計是否合理）。

WendyYehhh commented:
@monero778176 
Hi~
針對這個 PR，如果你願意的話，請在本週五（5/16）前 至少修改以下兩點，並在完成後通知我~ 謝謝！
1. GET 不要 return Hello World ，要提供 GET method 最好切分成另一個 endpoint 。
2.  `/api/view/order.py` 裡面關於 decorator  的實作是錯誤的。

monero778176 commented:
好的，謝謝給予的回饋，我會盡速修正。

monero778176 committed:
add ERD

monero778176 committed:
修正內容

補充關聯資料庫 ER Diagram
修正 import_order 中的 decorator  功能
新增 models 可能有關操作 api endpoint
新增對應 views
新增 views API 功能的 unittest

monero778176 committed:
Delete 快速刪除資料表.md

monero778176 committed:
Delete model設計.md

monero778176 committed:
Delete ERD_models.vuerd.json

monero778176 commented:
@WendyYehhh 您好:

已將上述的建議進行修正。
